[FIX] Fix of inconsistent indices #1726 #1731 #1732 #1734

dchigarev opened the pull request:
<!--
Thank you for your contribution! 
Please review the contributing docs: https://modin.readthedocs.io/en/latest/contributing.html
if you have questions about contributing.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #1726, #1731, #1732 and #1734 <!-- issue must be created for each patch -->
- [x] tests added and passing

Initially, this PR was fixing that `to_pandas` in pandas backend does not consider information about indices/columns of modin frame, only from partitions. But after that fix many tests started to fail because `df_equals`, that uses in many tests, use `to_pandas` to convert modin frame to pandas and then compare it with an expected result. It revealed that many operations returns modin frame with incorrect indices, so this PR is also fixing all of them.

dchigarev commented:
All check seems to be pass, except few that canceled because of timeout

dchigarev commented on modin/engines/ray/generic/io.py:
```diff
@@ -41,6 +41,6 @@ def func(df):
             return pandas.DataFrame()
 
-        result = qc._modin_frame._fold_reduce(1, func)
+        result = qc._modin_frame._apply_full_axis(1, func)
```
Look at the `func` above that we pass here to `_fold_reduce`, it returns empty DataFrame, so the shape partitions changing (including indices and columns). `_fold_reduce` can't handle changing shape on both axis, when `_apply_full_axis` can.

YarShev replied:
Okay, I see. Since we know that the result will be consist of empty partitions, may be, make sense pass to `_apply_full_axis` some values `new_index` and `new_columns` which are applied for empty dataframe?

dchigarev replied:
done

dchigarev commented on modin/backends/pandas/query_compiler.py:
```diff
@@ -480,7 +480,9 @@ def is_monotonic_reduce(df):
             return [common_case and edge_case]
 
-        return MapReduceFunction.register(is_monotonic_map, is_monotonic_reduce)(self)
+        return MapReduceFunction.register(
```
good note, but I suppose that I should fix this in a separated PR

YarShev replied:
I don't think it will take a lot of time. If it's convenient to do it in another PR, I don't mind.

dchigarev replied:
fixed it in this PR

YarShev replied:
Thanks!

dchigarev commented on modin/engines/base/frame/data.py:
```diff
@@ -1353,7 +1357,8 @@ def to_pandas(self):
             else:
                 df = pandas.DataFrame(columns=self.columns, index=self.index)
-        df.index.name = self.index.name
+        df.index = self.index
```
yeah, good point, fixed it in the last commit

dchigarev commented on modin/data_management/functions/binary_function.py:
```diff
@@ -39,6 +39,7 @@ def caller(query_compiler, other, *args, **kwargs):
                             axis,
                             lambda l, r: func(l, r.squeeze(), *args, **kwargs),
                             other._modin_frame,
+                            preserve_labels=False,
```
fixed in the last commit

dchigarev commented on modin/data_management/functions/groupby_function.py:
```diff
@@ -115,12 +115,14 @@ def compute_reduce(df):
-            else:
-                new_index = query_compiler.index
-                new_columns = None
+            new_index = None
```
done

YarShev commented:
LGTM! Let's wait for @devin-petersohn , and @gshimansky .

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -41,6 +41,6 @@ def func(df):
             return pandas.DataFrame()
 
-        result = qc._modin_frame._fold_reduce(1, func)
+        result = qc._modin_frame._apply_full_axis(1, func, new_index=[], new_columns=[])
```
Why is `new_index=[]` and not `None`? Same question for `new_columns`

dchigarev replied:
We know that the result of applying `func` will be an empty DataFrame, so instead of extra twice calling `get_indices` in `_apply_full_axis` we specify new indices explicitly here

devin-petersohn commented on modin/pandas/groupby.py:
```diff
@@ -295,7 +295,7 @@ def __getitem__(self, key):
                 and self._drop
             ):
-                key = [key] + list(self._by.columns)
+                key = list(self._by.columns) + [key]
```
Why is this change needed? It should not impact anything from what I see.

dchigarev replied:
this is the order in which pandas places columns at `groupby.__getitem__`, look at this reproducer:
```python
if __name__ == "__main__":
    import modin.pandas as pd
    import pandas
    import numpy as np

    data = {
        "col1": [0, 1, 2, 3],
        "col2": [4, 5, np.NaN, 7],
        "col3": [np.NaN, np.NaN, 12, 10],
        "col4": [17, 13, 16, 15],
        "col5": [-4, -5, -6, -7],
    }

    pd_grp = pandas.DataFrame(data).groupby("col4", as_index=False)
    md_grp = pd.DataFrame(data).groupby("col4", as_index=False)

    print("pd_columns:", pd_grp["col1"].count().columns)   # pd_columns: Index(['col4', 'col1'], dtype='object')
    print("md_columns:", md_grp["col1"].count().columns)  # md_columns: Index(['col1', 'col4'], dtype='object') 
```

devin-petersohn replied:
Yes, I see.

devin-petersohn commented on modin/engines/base/frame/data.py:
```diff
@@ -1353,7 +1357,9 @@ def to_pandas(self):
             else:
                 df = pandas.DataFrame(columns=self.columns, index=self.index)
-        df.index.name = self.index.name
+        else:
```
We should fix the columns/index internally for these cases instead of making this change.

We should also add an `assert` statement here.

dchigarev replied:
Look at the original reproducer of #1726, even if partitions contains correct indices, `concat` that used in `to_pandas` will lose information about levels, so I believe we should add `assert` statement for equality internal and external indices values and after that do that assignment
```python
df.index = self.index
df.columns = self.columns
```

dchigarev replied:
added assert check in the last commit

devin-petersohn commented on modin/data_management/functions/groupby_function.py:
```diff
@@ -115,19 +115,9 @@ def compute_reduce(df):
-                _reduce,
-                new_columns=new_columns,
-                new_index=new_index,
+                axis, by._modin_frame, _map, _reduce
```
Can we set `new_columns` and `new_index` based on some condition?

dchigarev replied:
Since we don't know what passed function would do with our DataFrame I'm afraid not. Let's remember our approach from #1645 where applied function transforms and changes shape of DataFrames in partitions. If we allow to apply functions like this, it's impossible to predict new indices/columns before computing.

dchigarev commented:
@devin-petersohn I guess I addressed all your comments, please take a look. I've also added propagation indices names like `["__reduced__"]` for Series into partitions, to it successfully passed assert check in `to_pandas`

devin-petersohn commented on modin/engines/base/frame/data.py:
```diff
@@ -74,6 +78,8 @@ def __init__(
         self._column_widths_cache = column_widths
         self._dtypes = dtypes
         self._filter_empties()
+        if validate_axes is not None and validate_axes is not False:
```
Can `validate_axes` be `None`? We can just give a type hint in the method signature for `validate_axes: bool = False`.

dchigarev replied:
Yeah, you're right, `validate_axes` is not supposed to be `None`, added type hint here

devin-petersohn commented on modin/engines/base/frame/data.py:
```diff
@@ -234,6 +246,64 @@ def _filter_empties(self):
         self._column_widths_cache = [w for w in self._column_widths if w > 0]
         self._row_lengths_cache = [r for r in self._row_lengths if r > 0]
 
+    def _validate_axis_equality(self, axis, update):
```
Can we use type hints and add some method level documentation as well?

dchigarev replied:
added docstring and type hints

devin-petersohn commented on modin/engines/base/frame/data.py:
```diff
@@ -234,6 +246,64 @@ def _filter_empties(self):
+        """
+        Validates and optionally updates internal and external indices
+        of modin_frame in specified mode. There is 3 modes supported:
+            1. "reduced only" - validates and updates indices on that axes
```
Instead of `"reduced only"` can we use `"reduced"` or something similar? (just a single word is preferred).

dchigarev replied:
changed it to `"reduced"`

dchigarev commented:
@devin-petersohn comments addressed and CI is green (for the exception of `black` check that seems to be frozen)

devin-petersohn commented on modin/engines/base/frame/data.py:
```diff
@@ -204,6 +211,21 @@ def _set_columns(self, new_columns):
                 self._dtypes.index = new_columns
         self._apply_index_objs(axis=1)
 
+    def _set_axis(self, axis, new_axis):
```
I prefer not to add this, you can use `self.axes[axis] = new_axis`

dchigarev replied:
`self.axes` will return us `[self.index, self.columns]`, roughly say just an array of values, so assignment like `self.axes[axis] = new_axis` will not do any effect, look at that reproduce:
```python
import pandas

df = pandas.DataFrame([1,2,3])
print(df.axes[0]) # RangeIndex(start=0, stop=3, step=1)
df.axes[0] = pandas.Index(['a', 'b', 'c'])
print(df.axes[0]) # RangeIndex(start=0, stop=3, step=1)
```

devin-petersohn replied:
Not enough coffee yet 😄

devin-petersohn commented on modin/engines/base/frame/data.py:
```diff
@@ -234,6 +256,74 @@ def _filter_empties(self):
+            axis, self._partitions, lambda df: df.axes[axis]
+        )
+        is_equals = self.axes[axis].equals(internal_axis)
+        if not update:
```
I did not see this getting used in this PR, how are you expecting this to be used?

devin-petersohn replied:
It may be better to just leave the `update` parameter out because I don't think we ever want to throw the internal error. What do you think?

dchigarev replied:
well, if you think so

removed that parameter in the last commit

dchigarev commented:
@devin-petersohn I've found out that `df_equals` [sorts columns labels](https://github.com/modin-project/modin/blob/master/modin/pandas/test/utils.py#L458-L466) before asserting frame equality, so after I've added assertion for columns equality (without that sorting) many tests appear to be failed. Are we really want to sort columns to check frames equality or this is a legacy stuff?

devin-petersohn commented:
>Are we really want to sort columns to check frames equality or this is a legacy stuff?

This is legacy, we should not be sorting. This is a different issue, let's open a new issue and PR for that and add the check in `df_equals` there. 

It looks like CI is failing to download the packages.

dchigarev commented:
> > Are we really want to sort columns to check frames equality or this is a legacy stuff?
> 
> This is legacy, we should not be sorting. This is a different issue, let's open a new issue and PR for that and add the check in `df_equals` there.
> 
> It looks like CI is failing to download the packages.

removed asserting columns from that branch and created a separate issue about sorting labels (#1784)

dchigarev commented:
@devin-petersohn CI is green, I guess PR is ready to merge
